### ISSUE TITLE
Clarify Fine tuning with HuggingFace

### ISSUE THREAD

**Issue by benemead** [zjunlp/MolGen#20]
Hi there,

Thank you for doing this work and releasing such an interesting model!

I'm struggling to follow the docs to perform my own fine tuning of molgen 7b from hugging face. Because the model is quantized I have to add a layer and use LORA (or similar). Since this isn't mentioned in any of the docs, wanted to clarify that this is your intended behavior?

I would also appreciate any advice you have on how I should structure my training datasets for property-based fine tuning?

Finally, if I simply provide a large set of SELFIES representing a chemical domain for fine tuning, is it reasonable to expect that tuned model will favor generation from that domain?

Thanks again!

**Comment by Fangyinfff** [comment]
Hi,

Thank you for your interest in our work.

**1. Regarding Quantization:**
The MolGen-7B model is not quantized. If you do not want the model to load in a quantized format, simply remove the load_in_8bit=True parameter. For example:
```
model = LlamaForCausalLM.from_pretrained("zjunlp/MolGen-7b", device_map="auto")
```

**2.	Training Dataset Structure for Property-Based Fine-Tuning:**
While the structure of your dataset largely depends on the specific task you want to achieve, you can consider referencing the approach we used in this repository to construct datasets.

**3.	Using SELFIES for Domain-Specific Fine-Tuning:**
Yes, providing a large set of SELFIES representing a specific chemical domain for fine-tuning will encourage the model to favor generation within that domain. Fine-tuning on domain-specific data adjusts the model’s output distribution, allowing it to generate molecules that better align with the input domain.
To achieve optimal results, consider:
	•	Ensuring high-quality and diverse training data.
	•	Validating outputs for domain-specific properties (e.g., molecule validity, uniqueness, and chemical relevance).

Best.

**Comment by zxlzr** [comment]
hi buddy, do you have any further questions?

**Comment by benemead** [comment]
Thank you @Fangyinfff and @zxlzr for such a prompt reply!

Re: 1 - yes I realized that right after I asked, my bad!

Re: 2 could you point me to a specific example in the repo of a dataset with structures and properties? I'm also curious about how the model uses property information when it's doing sequence completion

Re: 3 thank you for clarifying! Makes sense.

**Comment by Fangyinfff** [comment]
Please refer to the details in our paper, the data we provided, and the responses under this issue:
https://github.com/zjunlp/MolGen/issues/13#issuecomment-2219959522

**Comment by benemead** [comment]
Thank you!

Upon a closer read of both your paper and the referenced code, I think I'm a bit closer to understanding. 

Is below a correct summary?

For a source molecule you generate multiple candidates with the pre trained model, score those with the property of interest, and then pass the source, candidate, and candidate score as a form of semi supervised fine tuning?

If that's correct, is there a straightforward way to perform fine tuning in this manner via the transformers package and the hugging face model (I'm still learning the ins and outs of transformers and haven't seen this way described before)?

**Comment by Fangyinfff** [comment]
Hi,

For a given source molecule, the pre-trained model generates multiple candidate molecules, which are then scored based on the property of interest. Each candidate molecule is associated with a generation probability, and a rank-loss arises from the discrepancy between the ranking of these probabilities and the ranking of the property scores.

Currently, there isn’t a straightforward way to implement this using the transformers package directly. Custom modifications to the model’s loss function are necessary to incorporate this rank-loss.

**Comment by benemead** [comment]
Awesome, this is more or less exactly what I'd like to do (and IMO a really clever approach)! Any pointers on how I might approach a custom loss function with transformers and your model?

**Comment by Fangyinfff** [comment]
Hi,

I believe I’ve explained this clearly. You’ll need to modify the model you’re using by following the way the loss function (e.g. https://github.com/zjunlp/MolGen/blob/main/MolGen/src/utils.py#L176C1-L202C21) is constructed in my code. I think it would be best for you to understand the logic behind the code.  This will help you modify the model and construct the custom loss function (rank-loss) effectively.

**Comment by zxlzr** [comment]
Hi, do you have any further questions?

**Comment by benemead** [comment]
Nope, all set, thank you!
